### ISSUE TITLE
Development of Ubuntu PPA 

### ISSUE THREAD

**Issue by Robinlovelace** [abraunegg/onedrive#853]
I'm an Ubuntu user and read the [warning](https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md#important-information-for-ubuntu-users) in INSTALL.md. Happy to take a look at https://itsfoss.com/ppa-guide/ although I've never built Ubuntu pkgs before. Thanks for a great piece of software and pointers to get the latest version up.

**Comment by abraunegg** [comment]
@Robinlovelace 
Thanks - I cant give any guidance however as I do not use Ubuntu.

**Comment by Robinlovelace** [comment]
Had a look but it's bamboozling to me so unassigned, don't want to get ppls hopes up.

**Comment by yann1cks** [comment]
I had the same idea as @Robinlovelace and proudly present the first PPA I ever build:
https://launchpad.net/~yann1ck/+archive/ubuntu/onedrive-abraunegg

Its basically a rebuild of norbusan's Debian sid package, thats much easier than building from the scratch. I tried it on a clean Ubuntu 19.10 and it works fine. Give me feedback, if you encounter any problems.

**Comment by abraunegg** [comment]
@yann1cks 
Whilst I do not use Ubuntu, I cannot test - thanks for contributing.

The only suggestion I have is call the PPA / package 'onedrive' so that people do not have to work out what to install, or if they install 'onedrive', they get the older version.

The only reason why Arch Linux AUR uses 'onedrive-abraunegg' is that the current maintainer for the AUR 'onedrive' package does not want to rebase his packages (even after 2 years ..)

**Comment by yann1cks** [comment]
@abraunegg I thought that way the distinction between PPA and official package is clear. I dont know what the "best practice" is. But the more I think about is just "onedrive" is the way to go. Will change it later.

**Comment by Robinlovelace** [comment]
I think it will just pick the more recent (your) version of the package if your ppa is added.

**Comment by Robinlovelace** [comment]
That works fine and is common, enabling users to get close to the cutting edge (or use out-of-date packages in the default repos more commonly ; )

**Comment by mikejac** [comment]
Hmmm:
```$ sudo add-apt-repository ppa:yann1ck/onedrive-abraunegg
 This package provides an unofficial command line client for synchronization with Microsoft OneDrive. The client supports OneDrive Personal, OneDrive for Business, OneDrive for Office365 and Sharepoint. It supports one-way and two-way sync capabilities and securely connects to Microsoft OneDrive services.

Configuration and Usage: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md
 More info: https://launchpad.net/~yann1ck/+archive/ubuntu/onedrive-abraunegg
Press [ENTER] to continue or Ctrl-c to cancel adding it.

Hit:1 http://ppa.launchpad.net/mongoose-os/mos/ubuntu bionic InRelease
Hit:2 http://eu-frankfurt-1-ad-3.clouds.archive.ubuntu.com/ubuntu bionic InRelease
Hit:3 http://eu-frankfurt-1-ad-3.clouds.archive.ubuntu.com/ubuntu bionic-updates InRelease
Hit:4 http://eu-frankfurt-1-ad-3.clouds.archive.ubuntu.com/ubuntu bionic-backports InRelease
Ign:5 http://ppa.launchpad.net/yann1ck/onedrive-abraunegg/ubuntu bionic InRelease
Hit:6 http://security.ubuntu.com/ubuntu bionic-security InRelease
Err:7 http://ppa.launchpad.net/yann1ck/onedrive-abraunegg/ubuntu bionic Release
  404  Not Found [IP: 91.189.95.83 80]
Hit:8 https://download.docker.com/linux/ubuntu bionic InRelease
Reading package lists... Done
E: The repository 'http://ppa.launchpad.net/yann1ck/onedrive-abraunegg/ubuntu bionic Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.```

**Comment by mikejac** [comment]
Ubuntu 18.04

**Comment by yann1cks** [comment]
I changed the name of the package. 

@mikejac I tried to build a 18.04 package, but debbuilder and the d-compiler in LTS are outdated. I lack the experience to handle that situation, so I am not able to support the 18..04. If someone could hint me in the right direction I would give it another try.

**Comment by abraunegg** [comment]
@yann1cks 
> I tried to build a 18.04 package, but debbuilder and the d-compiler in LTS are outdated

Follow the directions in the [install guide] (https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md#dependencies-ubuntu-18x-ubuntu-19x--debian-9-debian-10---x86_64) for Ubuntu. I dont know if it is possible to do that, then build the package.

**Comment by abraunegg** [comment]
@yann1cks 
Also - it must be noted that:

* For older Ubuntu systems (xenial), you need to use a different compiler
* For Ubuntu i386 (bionic) you need to change the default linker, and use DMD as per web install
* For Ubuntu x86_64 (bionic and beyond) use DMD as per web install

Your PPA build script should be developed to use the right compiler / linker based on the details in the install guide if you can do this - that way, you can build the packages for each platform without issue, using the right toolset for the Operating System version & architecture.

Right now - your builds for i386 & x86_64 (bionic) are failing because of this reason.

The PPA would have to cover off:
* xenial
* bionic
* eoan
* focal

With the following architectures where possible:
* i386
* x86_64
* armhf
* arm64

Hope this helps

**Comment by shl-cx** [comment]
@yann1cks 
Hi,
when I executed this installation, the error feedback:
```
$ sudo add-apt-repository ppa:yann1ck/onedrive
 This package provides an unofficial command line client for synchronization with Microsoft OneDrive. The client supports OneDrive Personal, OneDrive for Business, OneDrive for Office365 and Sharepoint. It supports one-way and two-way sync capabilities and securely connects to Microsoft OneDrive services.

Configuration and Usage: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md
 More info: https://launchpad.net/~yann1ck/+archive/ubuntu/onedrive
Press [ENTER] to continue or Ctrl-c to cancel adding it.

Get:1 http://mirrors.digitalocean.com/ubuntu bionic InRelease [242 kB]
Hit:2 https://download.docker.com/linux/ubuntu bionic InRelease
Hit:3 http://mirrors.digitalocean.com/ubuntu bionic-updates InRelease
Hit:4 http://mirrors.digitalocean.com/ubuntu bionic-backports InRelease
Hit:5 http://security.ubuntu.com/ubuntu bionic-security InRelease
Hit:6 http://ppa.launchpad.net/yann1ck/onedrive/ubuntu bionic InRelease
Ign:7 https://netcologne.dl.sourceforge.net/project/d-apt d-apt InRelease
Get:8 https://netcologne.dl.sourceforge.net/project/d-apt d-apt Release [4357 B]
Get:9 https://netcologne.dl.sourceforge.net/project/d-apt d-apt Release.gpg [484 B]
Ign:9 https://netcologne.dl.sourceforge.net/project/d-apt d-apt Release.gpg
Reading package lists... Done
W: GPG error: https://netcologne.dl.sourceforge.net/project/d-apt d-apt Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY EBCF975E5BA24D5E
E: The repository 'https://netcologne.dl.sourceforge.net/project/d-apt d-apt Release' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```
Since the operation of updating repo **d-apt** is insecure, the installation can't be done.
So here is A easy way to fix it: (from [d-apt documentation](https://d-apt.sourceforge.io/))
_I suppose adding signing key is another solution, not try yet._
To enable it, add the repository sources:

```
$ sudo wget https://netcologne.dl.sourceforge.net/project/d-apt/files/d-apt.list -O /etc/apt/sources.list.d/d-apt.list
```
then update local info and install "d-apt" public key (fingerprint 0xEBCF975E5BA24D5E):
```
$ sudo apt-get update --allow-insecure-repositories && sudo apt-get -y --allow-unauthenticated install --reinstall d-apt-keyring && sudo apt-get update
```

then, continue the installation:
```
$ sudo add-apt-repository ppa:yann1ck/onedrive
 This package provides an unofficial command line client for synchronization with Microsoft OneDrive. The client supports OneDrive Personal, OneDrive for Business, OneDrive for Office365 and Sharepoint. It supports one-way and two-way sync capabilities and securely connects to Microsoft OneDrive services.

Configuration and Usage: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md
 More info: https://launchpad.net/~yann1ck/+archive/ubuntu/onedrive
Press [ENTER] to continue or Ctrl-c to cancel adding it.

Get:1 http://mirrors.digitalocean.com/ubuntu bionic InRelease [242 kB]
Hit:2 http://mirrors.digitalocean.com/ubuntu bionic-updates InRelease
Hit:3 http://mirrors.digitalocean.com/ubuntu bionic-backports InRelease
Hit:4 https://download.docker.com/linux/ubuntu bionic InRelease
Hit:5 http://security.ubuntu.com/ubuntu bionic-security InRelease
Hit:6 http://ppa.launchpad.net/yann1ck/onedrive/ubuntu bionic InRelease
Ign:7 https://netcologne.dl.sourceforge.net/project/d-apt d-apt InRelease
Hit:8 https://netcologne.dl.sourceforge.net/project/d-apt d-apt Release
Fetched 242 kB in 1s (192 kB/s)
Reading package lists... Done
```
As the output shows the success of the installation.
**BUT** the problem following:
`sudo apt-get onedrive`
After this, I still got the old version of Onedrive, what should I do exactly to install the abraunegg-onedrive.

I'd be appreciate if u can reply asap.

**Comment by abraunegg** [comment]
@shl-cx 
You should not be using or installing the 'd-apt' repo - that provides a very out of date D Compiler. Once this client is packaged, that is a binary that does not need building - thus, you should not need that 'd-apt' repo. I do not use Ubunutu, so I would remove it any reference to it - I cannot advise on that.

I do suggest that at this time until packaging is done for the entire spectrum and that the installation instructions are updated to include directions for using a PPA, try to use the PPA and if that fails (like it is for you) build from source using the directions provided for Ubuntu.

**Comment by shl-cx** [comment]
> @shl-cx
> You should not be using or installing the 'd-apt' repo - that provides a very out of date D Compiler. Once this client is packaged, that is a binary that does not need building - thus, you should not need that 'd-apt' repo. I do not use Ubunutu, so I would remove it any reference to it - I cannot advise on that.
> 
> I do suggest that at this time until packaging is done for the entire spectrum and that the installation instructions are updated to include directions for using a PPA, try to use the PPA and if that fails (like it is for you) build from source using the directions provided for Ubuntu.

Thank you for your help.

I did use PPA to install but failed, whereas the ubuntu cannot ship with the latest abraunegg-onedrive.
idk what happened, here are my steps:

`sudo add-apt-repository ppa:yann1ck/onedrive`
and let's see the `sources.list.d`
```
$ tree /etc/apt/sources.list.d
/etc/apt/sources.list.d
├── d-apt.list
├── d-apt.list.save
├── docker.list
├── docker.list.save
├── yann1ck-ubuntu-onedrive-bionic.list
└── yann1ck-ubuntu-onedrive-bionic.list.save
```
where the yann1ck-ubuntu-onedrive is added.
add key:
```
sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys E2FB3A8B1D64A13B844479700512B5FE7F557AB1 
```
so last steps are update&install
`sudo apt-get update` and `sudo apt-get install onedrive`
but after installation, check the version of it:
```
$ onedrive --version
onedrive v1.0.1
```
so sad I don't know which step I missed or mistake, hope u can help me.

**Comment by abraunegg** [comment]
@shl-cx 
Firstly I would remove:
* d-apt.list
* d-apt.list.save

These repo's are not needed .. adding this comes from 'skilion' repo instructions which are incorrect. If you have followed any of these instructions, you need to back them out - including uninstalling any packages: https://github.com/skilion/onedrive#dependencies-ubuntudebian

You then probably have to 'configure' your system to prefer to use the PPA rather that Ubuntu sources for 'onedrive' ... again ... I do not use Ubuntu. I cannot you assist further.

Suggest if the above is not possible, compile from source.

**Comment by shl-cx** [comment]
Well, I tried all means but still failed.

So I post my steps here, for anyone could see it and plz announce if you guys find out the solution.

I commented out all ubuntu official sources except the PPA (artificially modify)
> I don't compile it coz my Ubuntu server shipped with less than 1G RAM, perhaps I will try to compile on other devices and help it'll work.
```
# the content of sources.list
deb http://ppa.launchpad.net/yann1ck/onedrive/ubuntu bionic main
deb-src http://ppa.launchpad.net/yann1ck/onedrive/ubuntu bionic main
```
Update..
```
$ sudo apt-get update
Hit:1 https://download.docker.com/linux/ubuntu bionic InRelease
Hit:2 http://ppa.launchpad.net/yann1ck/onedrive/ubuntu bionic InRelease
Reading package lists... Done
```
However, it still can't get the updated version from PPA

```
$ sudo apt-get install onedrive=2.4.0-4~eoan
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package onedrive
```

So you see, it still cannot fetch the pkg from PPA.. (FYI, nothing changed after reboot)

**Comment by abraunegg** [comment]
@shl-cx 
I am sorry I cannot assist any further ... I do not use Ubuntu.

It is best that you build from source following these directions:
* https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md#dependencies-ubuntu-18x-ubuntu-19x--debian-9-debian-10---x86_64
* https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md#compilation--installation

**Comment by ToucheSir** [comment]
I just experienced the same installation issues as @shl-cx. After looking into it, it appears that there are no successfully built packages for Ubuntu 18.04 in @yann1cks PPA. In other words, apt finds no packages in the PPA and falls back to the default out-of-date version.

See the "source" section under https://launchpad.net/~yann1ck/+archive/ubuntu/onedrive/+packages and https://launchpad.net/~yann1ck/+archive/ubuntu/onedrive/+build/19114118 for the `amd64` build in particular. Here's the relevant snippet from the failed build log:
```
------------------------------------------------------------------------------+
| Install onedrive build dependencies (apt-based resolver)                     |
+------------------------------------------------------------------------------+

Installing build dependencies
Reading package lists...
Building dependency tree...
Reading state information...
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 sbuild-build-depends-onedrive-dummy : Depends: debhelper-compat (= 12)
E: Unable to correct problems, you have held broken packages.
apt-get failed.
Package installation failed
Reading package lists...
Building dependency tree...
Reading state information...
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Reading package lists...
Building dependency tree...
Reading state information...
The following packages will be REMOVED:
  sbuild-build-depends-core-dummy*
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
After this operation, 0 B of additional disk space will be used.
(Reading database ... 12532 files and directories currently installed.)
Removing sbuild-build-depends-core-dummy (0.invalid.0) ...

+------------------------------------------------------------------------------+
| Cleanup                                                                      |
+------------------------------------------------------------------------------+

Purging /<<BUILDDIR>>
Not removing build depends: as requested
E: Package build dependencies not satisfied; skipping
```

**Comment by yann1cks** [comment]
As stated two weeks ago, the 18.04 package is not working. @ToucheSir shows that there is a problem with debhelper. There is also one with the d-compiler. Thats also the reason, that @shl-cx system doesnt automatically use the PPA.

At the moment I have no time to fix the issue, but I will try to do it in the next week.

**Comment by mikestr** [comment]
Hi @yann1cks 
I have been using onedrive for a while now, and am running 18.04.  Maybe I can help you with the PPA? Bit of a noob on PPAs, but maybe we can get it tested and working together.

**Comment by abraunegg** [comment]
@yann1cks 
Any update on the PPA development?

**Comment by yann1cks** [comment]
Not yet, I want to build a version 20.04 on the weekend and then I will look into 18.04.

**Comment by yann1cks** [comment]
I have an update. I managed to automated the building of the PPA on Launchpad and the build for 20.04 seems to be working fine. The building process simply uses norbusan's debian repository and automatically creates a PPA.

The bad news is that I did not manage to build a version for Ubuntu 18.04 (bionic). The main reason is the [outdated version of ldc](https://packages.ubuntu.com/de/bionic/ldc) in the bionic repositories. Bionic comes with version 1.8 but version 1.12 is required. Launchpad also does not allow the use of external sources, like dlang.org. Backporting to an earlier ldc version is also not an option for me because I expect it to be too complicated.

I see several options now. Please let me know if I miss another possible option:
- The PPA will not work with Ubuntu 18.04.
- I could create an ldc PPA, which I could in turn use as dependency for the onedrive PPA. Quite hacky and I am not sure that this actually works.
- Building a Snap or AppImage, the later should be possible on the OpenSUSE build service. I have no experience with Snap or Appimages.

**Comment by mikestr** [comment]
what about using dmd instead?  Would that help things?

**Comment by PetarKirov** [comment]
Regarding Ubuntu, are snaps allowed to be used as part of the build process? The dmd and ldc snaps are more up to date there: 
- https://snapcraft.io/ldc2
- https://snapcraft.io/dmd

**Comment by athosbr99** [comment]
Just to add up, I am using the PPA from @yann1cks on Ubuntu 20.04 and so far it is working fine.

**Comment by yann1cks** [comment]
@mikestr dmd is not in the Ubuntu repositories. I think the reason are some closed source components. So i cannot use it. @PetarKirov Its the same with snaps. PPAs are basically debain packages so there is no way to integrate snaps in the build process on Launchpad.

**Comment by ffes** [comment]
FYI: on the 20.04 (focal) [v2.3.13](https://packages.ubuntu.com/focal/onedrive) is included, and on 19.10 (eaon) it is [v2.3.9](https://packages.ubuntu.com/eaon/onedrive).

Takes nothing away from having a PPA that ensures the availability of the latest version, but in a newer Ubuntu you will get a relative recent version.

**Comment by abraunegg** [comment]
@ffes 
> FYI: on the 20.04 (focal) [v2.3.13](https://packages.ubuntu.com/focal/onedrive) is included, and on 19.10 (eaon) it is [v2.3.9](https://packages.ubuntu.com/eaon/onedrive).

And this is precisely the reason why this warning exists: https://github.com/abraunegg/onedrive/blob/master/docs/INSTALL.md#important-information-for-ubuntu-users

v2.3.9 is from September 2019. Significant development has been done + significant bugs have been fixed

v2.3.13 is from December 31 2019. It is probably the most stable of all the v2.3.x series, but also significant bugs were found and fixed between that release and even now (~5 months)

In short - if you are on Ubuntu .. **dont** use the packages. If you care about fixing this - become the Ubuntu Package Maintainer and update the Ubuntu packages so that they are kept current - otherwise assist in developing a PPA or build from source.

**Comment by abraunegg** [comment]
@yann1cks 
Any update on developing the PPA and being compatible with 18.x Ubuntu releases?

**Comment by yann1cks** [comment]
@abraunegg I will keep supporting the newer Ubuntu versions of Ubuntu in the PPA  but I wont be able to support 18.04, because of the ldc version in this Ubuntu version (1.8 vs 1.12). dmd cannot be used because it is not part of the Ubuntu repos. An alternative to install newer versions of ldc or dmd would by using the Snap packages, but Snaps cannot be used as dependencies. So that is no solution. Therefore backporting would be necessary and I dont have the time or the experience to do this. 

Another option, and perhaps a better one, would be a Snap package (or an AppImage), but I have not experience with these formats.

**Comment by luni3359** [comment]
Does this packaged version support notifications?

**Comment by abraunegg** [comment]
@luni3359
Please do not respond to closed threads.
